### ISSUE TITLE
Trigger load state

### ISSUE THREAD

**Issue by andykirk** [DataTables/DataTables#630]
Hi,
I've seen this sort of issue crop up on the old forum, but nowhere recently, so I thought I'd present my use case since none of the others seemed to make it clear_why_ this would be useful.

Scenario: I want to allow my users the option of saving the state of a particularly complicated table so they (or others) can load it again it in the future.
Saving the state to a json file was easy enough, but in order to load it again (in response to a button click) I need to be able to trigger the load state callback in the same way as `state.save()`.

My first hacky attempts at this seem to work:

~~~
_api_register( 'state.load()', function () {
    return this.iterator( 'table', function ( settings ) {
        _fnLoadState( settings, settings.oInit );
    } );
} );
~~~
(Following `_api_register( 'state.save()', function () {...}` ending on line 8866 of the main JS file).

I realise this isn't quite how the save/load state functionality was intended to be used so this on it's own isn't enough to add this functionality completely, but I've got it working well enough and I'm happy to share the relevant parts of the code if it helps my case or makes things clearer.

I'd be happy to discuss further if need be.

Thanks.

**Comment by DataTables** [comment]
Old forum? The [DataTables forum](//datatables.net/forums) is active and much preferred for support discussions over the github issues list (its proving to be difficult for me to trace both support channels at the same time).

State loading after initialisation time is an interesting idea, but not something I think I want to include in the DataTables core at the moment. It would be useful as a plug-in if you fancy submitting a PR to the [plug-ins repo](https://github.com/DataTables/Plugins/tree/master/api) (have a look at the `sum()` and `average()` files for the current way of doing plug-ins).

**Comment by andykirk** [comment]
Thanks for the quick response.

Apologies - I meant I only found similar requests for the legacy version of DataTables. I'll raise any future issues on the forum :-)

I did (briefly) look into creating a plugin but I couldn't find a way of properly calling `_fnLoadState` from with a plugin. My understanding is it's a private function. The closest I got was something like:
~~~
$.fn.dataTable.Api.register( 'state.load()', function () {
    return this.iterator( 'table', function (  ) {
        var s = $('#the_table').DataTable().settings()[0];
        $.fn.dataTable()._fnLoadState(s, s.oInit);
    } );
} );
~~~
But I get the error:
~~~
TypeError: settings is null
var columns = settings.aoColumns;
~~~
from the `_fnLoadState` function - line 4962 whatever I try. Any ideas / pointers on this? Is it possible to properly access the `_fnLoadState` from within a plugin.

As I say, adding the `state.load()` function to the core, then calling it (followed by `table.draw()`) does the job. I'm pretty new to DataTables so I haven't had chance dive into the docs and code as much as I'd like yet, but I thought it wise to ask first.

Thanks for such an amazing tool, by the way. :-)

**Comment by DataTables** [comment]
There is a nasty hack that I use to expose the private functions in the settings object - you can do `settings.oApi._fnLoadState`. Its not something I'm proud of, and don't encourage using (thus is isn't documented), but its the only want to expose the internal methods at the moment!

One day I'll get around to make it all instance based :-)

And thanks!

**Comment by andykirk** [comment]
Ok, thanks. I'd rather use a nasty hack than hack the core :-)

I'll give that a go and see how I get on.

FYI this is the forum post I found:
http://datatables.net/forums/discussion/21399/calling-fnstateload-manually

It mentions 'legacy' at the top, and you asked the OP "why would you want to call _fnLoadState?".
My plan outlined above provides an answer to that question. I do feel it's a shame there isn't a `state.load()` companion to `state.save()` so if there's ever anything I can do to help (/ help convince you), just ask :-)
Totally understand though.

Thanks again.

**Comment by DataTables** [comment]
Ah I see - that notice is shown for any post that references the legacy API. it isn't the forum that is legacy / old, but rather the content of the post might not be relevant to the current version of DataTables.

To convince me? A few more people asking for it! I isn't something that comes up all that frequently, so I don't think it is something that many want to use, and thus I want to try and keep the core reasonably slim (its already bigger than my goal, need to do something about that...!).

**Comment by andykirk** [comment]
Fair enough :-)

Here's an update:
~~~
jQuery.fn.dataTable.Api.register( 'state.load()', function () {        
    return this.iterator( 'table', function ( s ) {
        s.oApi._fnLoadState(s, s.oInit);
    } );
} );
~~~
Works like a charm. Can you see anything obviously wrong there (just in case)?

The whole thing works like this: (assuming the above is included as well)
~~~
var autosave_state = true;
var autoload_state = true;

// Omitted - api_url must correspond to your server-side code
// that handles your file loading and saving.
var api_url = 'your_api';

var table = jQuery('#your-table').DataTable({
    "stateSave": true,
    "stateLoadCallback": function (settings) {
        if (autoload_state) {
            console.log('Autoload - ignoring');
            // Or re-implement localStorage / implement other storage.
            // Not needed in my case.
            return;
        }
        var o;
        var filename = 'the_selected_filename';
        // E.g. get the filename from a list of files.
        
        $.ajax({
            'url': api_url + '?name=' + filename,
            'dataType': 'json',
            'async': false,
            'success': function(data){
                o = data;
            }
        });
        
        // Reset autoload_state state:
        autoload_state = true;
        return o;
    },
    "stateSaveCallback": function(settings, data){
        if (autosave_state) {
            console.log('Autosave - ignoring');
            // Or re-implement localStorage / implement other storage.
            // Not needed in my case.
            return;
        }

        var filename = 'user_input_filename';
        // E.g. get the filename from an input.
        
        $.ajax({
            'url': api_url + '?name=' + filename,
            'data': data,
            'dataType': 'json',
            'type': 'POST',
            'success': function(data){
                // Show a 'Saved' message or something.
            }
        });
        
        // Reset autosave_state state:
        saving_data = false;
    }
});
~~~
Hopefully this is useful to someone.

Thanks again.

**Comment by DataTables** [comment]
I don't immediately see anything wrong with it. The private function wasn't designed with that use in mind so it is possible that aspects of it might not work, but I'm not certain off the top of my head. Have you tested global and column based filtering?

**Comment by andykirk** [comment]
Yes - the `search:{}` and `columns[x]:{search:{}}` data that's loaded from the file is applied correctly (and search input repopulated), if that's what you mean?

I realise the private function wasn't intended for that use, but it seems to work great so I'll risk it. DataTables is brilliant, and triggering the `_fnLoadState` was the only snag I've hit so far. The 'nasty hack' is certainly better than the alternative in my book :-)

Cheers

**Comment by DataTables** [comment]
Sounds good to me :-). Nice hack!

**Comment by andykirk** [comment]
I thought it was too good to be true! The `_fnLoadState` call in the API plugin doesn't trigger column visibility change. I've tried everything I can think of, but I can't figure out what I need to call to make it happen.
If you have any ideas/pointers on how to do that, that would be great, but I suspect I may need to destroy the table and reinitialise? If that's the case, then I'll need to inject the saved state into the reinitialisation, but I'm not sure how to do that either.
Any help would be greatly appreciated.

Thanks.

**Comment by DataTables** [comment]
You'd probably need to run over the column visibility flags and set them using the `column().visible()` method. To be honest, that might actually be a more reliable way of doing the whole thing. Use the public API with the publicly documented state save object structure. It is more code, but more reliable and not using a private method.

**Comment by andykirk** [comment]
Understood. I'll try fixing just the columns thing for now, but I'll have a think about how I might best create a separate plugin that avoids using `_fnLoadState` entirely.

Thanks again for your quick repsonse and help.

**Comment by andykirk** [comment]
Hi, I tried using the public API to no avail. Something odd happens to the ability to set column visibility when the state is force-loaded.
In the end, I've got it working like this:
~~~
$.fn.dataTable.Api.register( 'state.load()', function () {
    return this.iterator( 'table', function ( s ) {
        // Force the state load. Note something isn't quite right with column visibility...
        s.oApi._fnLoadState(s, s.oInit);
        // ...so we need to check for which ones should be hidden in the loaded state...
        var hidden_cols = [];
        $.each(s.oLoadedState.columns, function(i, column) {
            if (!column.visible) {
                hidden_cols.push(i);
            }
        });
        // set them all to be visible...
        s.oInstance.DataTable().columns().visible(true);            
        // ...thne hide the ones that should be hidden:
        s.oInstance.DataTable().columns(hidden_cols).visible(false);
    });
});
~~~

Not ideal, but at least it works now.

Thanks again for all your help.

(Note I've updated the sample code above to include this workaround.)

**Comment by DataTables** [comment]
Reopening this as I need to consider if DataTables should provide this as part of the core - sync XML requests are being banned in new versions of the XMLHttpRequest documentation unless they are in a web worker, so DataTables needs to provide a way to allow state to be loaded async - see [this thread](https://www.datatables.net/forums/discussion/30403).

The other option is that a callback is provided for the state load callback.

**Comment by dpsk** [comment]
I'm not sure if this is the right place to describe my issue.
Case:
Ajax datatable, link that will replace content of the page dynamically(ajax).
As soon as link clicked we are pushing state change to History.
Click 'Back' button, previous page with datatable loaded as expected, ajax data is in place, but not the state of datatable(maybe it wasn't even initialised?).
Hit F5, page refreshed, state applied correctly.
TLDR: State not applied when the back button.
Hints?

**Comment by DataTables** [comment]
Support requests should be in the [DataTables forums](//datatables.net/forums) with a link to a test page showing the issue please.

**Comment by mrcrbrown** [comment]
+1 for this feature.

I'm  the @andykirk extension for a project where there are configurable "Views" of a 30+ column DataTable. User can change column visibility, order and filter and then commit those with a userable name. Each Name can then be selected and it rebuilds the DataTable. Generically I had problems with ColReorder and getting yadcf to update so I just added those project specifics to my implementation.

Thanks

**Comment by andykirk** [comment]
Hi @mrcrbrown - can you clarify your post above?
"the andykirk extension" I assume must be a mistake and sort of implies that we're the same person or somehow connected. I just would like to make sure anyone reading this that it's not the case.

Thanks.

**Comment by mrcrbrown** [comment]
haha yes I meant to write "I'm_** using **_the @andykirk extension"

**Comment by 719media** [comment]
Would like to see this feature implemented. I have table states being saved to a database, and a user can refresh a table by using these saved states, and it would be nice to have a built-in function for reloading a saved state

**Comment by DataTables** [comment]
Planned for inclusion in the next major version. (I'm not sure when that release will be though, before anyone asks :-) ).

**Comment by 719media** [comment]
I could not reliably get the code above to work, but it did get me on the right path. For some reason my table headers were not hiding on the visible calls when using that code, I'm not entirely sure why. Anyway, I did manage to get something else working, and am including it here in case it helps anyone else. I used some of the code that @andykirk wrote to get started on the right path. If you use the stateSave and Load callbacks to control your data, then you should be able to load data like so:

```javascript
function load_state(state_data) {
    if(state_data.hasOwnProperty('length')) {
        dt.page.len( state_data.length );
    }
    if(state_data.hasOwnProperty('ColReorder')) {
        dt.colReorder.order( state_data.ColReorder, true );
    }
    if(state_data.hasOwnProperty('order')) {
        dt.order(state_data.order);
    }
    if(state_data.hasOwnProperty('columns')) {
        var visible_cols = [];
        var hidden_cols = [];
        $.each(state_data.columns, function(i, column) {
            if (column.visible === false) {
                hidden_cols.push(i);
            } else {
                visible_cols.push(i);
            }
        });
        // set the visible ones...
        dt.columns(visible_cols).visible(true, false);
        // ...then hide the others:
        dt.columns(hidden_cols).visible(false, false);
    }
}
```

Keep in mind the ORDER that these are running is important (if you do the ColReorder AFTER the order/columns calls, you will get bad results as these calls are index based).

**Comment by farzher** [comment]
hi, can we trigger `stateLoadCallback` yet? i tried to hack it with no success

**Comment by DataTables** [comment]
No, not yet I'm afraid.

**Comment by Samistine** [comment]
Wanted
